src/clp_ffi_py/ir/native/PySerializer.cpp

```diff
@@ -0,0 +1,569 @@
+#include <clp_ffi_py/Python.hpp>  // Must always be included before any other header files
+
+#include "PySerializer.hpp"
+
+#include <array>
+#include <cstddef>
+#include <cstdint>
+#include <optional>
+#include <span>
+#include <type_traits>
+#include <utility>
+
+#include <clp/ffi/ir_stream/protocol_constants.hpp>
+
+#include <clp_ffi_py/api_decoration.hpp>
+#include <clp_ffi_py/error_messages.hpp>
+#include <clp_ffi_py/ir/native/error_messages.hpp>
+#include <clp_ffi_py/PyObjectCast.hpp>
+#include <clp_ffi_py/PyObjectUtils.hpp>
+#include <clp_ffi_py/utils.hpp>
+
+namespace clp_ffi_py::ir::native {
+namespace {
+/**
+ * Callback of `PySerializer`'s `__init__` method:
+ */
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays)
+PyDoc_STRVAR(
+        cPySerializerDoc,
+        "Serializer for CLP key-value pair IR stream.\n"
+        "This class serializes log events into CLP IR format (using four-byte-encoding) and writes"
+        " the serialized data to a specified byte stream object.\n\n"
+        "__init__(self, output_stream, buffer_size_limit)\n\n"
+        "Initializes a :class:`Serializer` instance with the given output stream. Notice that each"
+        " object should be strictly initialized only once. Double initialization will result in"
+        " memory leak.\n\n"
+        ":param output_stream: A writable byte stream to which the serializer will write serialized"
+        " IR byte sequence.\n"
+        ":param buffer_size_limit: The buffer size in bytes that will trigger the internal buffer"
+        " flush, defaults to 65536. The serialized log events will be first buffered inside the"
+        " internal buffer. When the size of the buffer exceeds this limit, the internal buffer will"
+        " be written to the output stream.\n"
+);
+CLP_FFI_PY_METHOD auto
+PySerializer_init(PySerializer* self, PyObject* args, PyObject* keywords) -> int;
+
+/**
+ * Callback of `PySerializer`'s `serialize_msgpack` method.
+ */
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-c-arrays)
+PyDoc_STRVAR(
+        cPySerializerSerializeMsgpackMapDoc,
+        "serialize_msgpack_map(self, msgpack_map)\n"
+        "--\n\n"
+        "Serializes the given msgpack byte sequence into as a log event into CLP IR format.\n\n"
+        ":param msgpack_map: A byte sequence encoded in msgpack as the input log event."
+        " The unpacked result must be a msgpack map with all keys as strings.\n"
+        ":return: The number of bytes serialized.\n"
+        ":raise IOError: If the serializer has already been closed.\n"
+        ":raise TypeError: If the unpacked result is not a msgpack map.\n"
+        ":raise RuntimeError: If it fails to unpack the given msgpack byte sequence, or the"
+        " serialization method returns failure.\n"
+);
+CLP_FFI_PY_METHOD auto
+PySerializer_serialize_msgpack_map(PySerializer* self, PyObject* msgpack_map) -> PyObject*;
```

Comment:
How about `PySerializer_serialize_log_event`?

Reply:
There are two ways to do this:
- We name it `serialize_log_event`, which allows users to input either msgpack bytes or Python dictionary and serialize it into IR format
- We can be explicit on the API naming, i.e., `serialize_msgapck_map`, `serialize_dict`

Reply:
How about `serialize_log_event_from_msgpack_map` and `serialize_log_event_from_dict`?

Reply:
ok, `from_msgpack_map` works for me

Reply:
changed to `serialize_log_event_from_msgpack_map`. Let's add `serialize_log_event_from_dict` in future PRs